praw/models/reddit/subreddit.py

```diff
@@ -2174,6 +2179,84 @@ def _upload_image(self, image_path, data):
                 raise APIException(error_type, error_value, None)
             return response
 
+    def _upload_style_asset(self, image_path, data):
+        data['filepath'] = basename(image_path)
+        data['mimetype'] = 'image/jpeg'
+        if image_path.lower().endswith('.png'):
+            data['mimetype'] = 'image/png'
+        url = API_PATH['style_asset_lease'].format(subreddit=self.subreddit)
+
+        # until we learn otherwise, assume this request always succeeds
+        upload_lease = self.subreddit._reddit.post(url, data=data)['s3UploadLease']
+        upload_data = {item['name']: item['value']
+                       for item in upload_lease['fields']}
+        upload_url = 'https:{}'.format(upload_lease['action'])
+
+        with open(image_path, 'rb') as image:
+            response = self.subreddit._reddit._core._requestor._http.post(
+                upload_url, data=upload_data, files={'file': image})
+        response.raise_for_status()
+
+        structured_style = {}
+        image_url = response.headers['location']
+        structured_style[data['imagetype']] = image_url
+
+        style_url = API_PATH['structured_styles'].format(subreddit=self.subreddit)
+        self.subreddit._reddit.patch(style_url, structured_style)
+        return response
```

Comment:
Because this method is named `_upload_style_asset`, I think it makes more sense to `return response.headers['location']` on (what is currently) line 2201 and perform these interactions with the structured_styles endpoint in the methods that call `_upload_style_asset` (such as `upload_banner`). In my view, one method should handle the file upload and return that file's URL, and other methods should use that result to add the banner images.

Reply:
Agreed.  _upload_style_asset will now return the S3 URL for the uploaded asset and it will be the responsibility of the calling function to figure out the right thing to do with that information.